platform/admin/configurator/admin-configurator-impl/src/main/java/org/codice/ddf/admin/configurator/impl/ConfigOperation.java

```diff
@@ -34,58 +45,244 @@
             }
             return ResultImpl.pass();
         }
 
         @Override
         public Result<Void> rollback() throws ConfiguratorException {
-            return ResultImpl.rollback();
+            try {
+                createConfig(configFile, strategy, originalConfigs);
+                return ResultImpl.rollback();
+            } catch (IOException e) {
+                return ResultImpl.rollbackFail(new ConfiguratorException("Error creating config: ",
+                        e));
+            }
+        }
+
+        @Override
+        public String toString() {
+            return "Deleting config " + configFile.getAbsolutePath();
+        }
+    }
+
+    /**
+     * Transactional handler for updating config files
+     */
+    private static class UpdateHandler extends ConfigOperation {
+        private final boolean keepIfNotPresent;
+
+        private Dictionary<String, Object> originalConfigs;
+
+        private UpdateHandler(Path configFile, Dictionary<String, Object> configs,
+                boolean keepIfNotPresent, PersistenceStrategy strategy) {
+            super(configFile, configs, strategy);
+            this.keepIfNotPresent = keepIfNotPresent;
+            this.originalConfigs = new Hashtable<>();
+        }
+
+        @Override
+        public Result<Void> commit() throws ConfiguratorException {
+            Dictionary<String, Object> configsToWrite = new Hashtable<>();
+            originalConfigs = readConfig(configFile, strategy);
+
+            if (keepIfNotPresent) {
+                putAll(originalConfigs, configsToWrite);
+            }
+            putAll(configs, configsToWrite);
+
+            try {
+                FileOutputStream outputStream = new FileOutputStream(configFile);
+                strategy.write(outputStream, configsToWrite);
+                return ResultImpl.pass();
+            } catch (IOException e) {
+                return ResultImpl.fail(new ConfiguratorException("Error writing config: ", e));
+            }
+        }
+
+        @Override
+        public Result<Void> rollback() throws ConfiguratorException {
+            try {
+                FileOutputStream outputStream = new FileOutputStream(configFile);
+                strategy.write(outputStream, originalConfigs);
+                return ResultImpl.rollback();
+            } catch (IOException e) {
+                return ResultImpl.rollbackFail(new ConfiguratorException("Error writing config: ",
+                        e));
+            }
+        }
+
+        @Override
+        public String toString() {
+            return "Updating config " + configFile.getAbsolutePath();
+        }
+
+        private static void putAll(Dictionary<String, Object> source,
+                Dictionary<String, Object> target) {
+            Enumeration keys = source.keys();
+            while (keys.hasMoreElements()) {
+                String key = (String) keys.nextElement();
```

Comment:
If you use Enumeration<String> keys = ... above, you shouldn't need the cast here.